hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -881,6 +909,10 @@ public String startCommit() {
    * Provides a new commit time for a write operation (insert/update/delete/insert_overwrite/insert_overwrite_table) with specified action.
    */
   public String startCommit(String actionType, HoodieTableMetaClient metaClient) {
+    if (needsUpgradeOrDowngrade(metaClient)) {
```

Comment:
the `startCommitXXX` methods don't create table upfront, instead use service client to first rollback, which fails without an upgrade done first.